File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        vector<int> graph[n];
        for (auto e : edges) {
            graph[e[0]].push_back(e[1]);
            graph[e[1]].push_back(e[0]);
        }
        queue<int> q;
        vector<bool> visited(n, false);
        q.push(source);
        visited[source] = true;
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            if (node == destination) {
                return true;
            }
            for (auto it : graph[node]) {
                if (!visited[it]) {
                    visited[it] = true;
                    q.push(it);
                }
            }
        }
        return false;
    }
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }
        boolean[] visited = new boolean[n];
        dfs(adj, source, visited);
        return visited[destination];
    }
    
    public void dfs(List<List<Integer>> adj, int src, boolean[] visited) {
        visited[src] = true;
        for (Integer neigh : adj.get(src)) {
            if (!visited[neigh]) {
                dfs(adj, neigh, visited);
            }
        }
    }
}